File: mrecord-java/src/main/java/com/dcz/mrecord/common/Result.java
package com.dcz.mrecord.common;

import lombok.Data;

/**
 * 响应体
 *
 * @author dcz
 * @since 2026/04/08
 */
@Data
public class Result<T> {
    /**
     * 响应码
     */
    private String code;

    /**
     * 响应消息
     */
    private String message;

    /**
     * 响应数据
     */
    private T data;

    /**
     * 私有化构造，禁止外部new
     */
    private Result() {
    }

    /**
     * 成功返回
     *
     * @param <T> 响应的数据类型
     * @return 响应数据
     */
    public static <T> Result<T> success() {
        return success(null);
    }

    /**
     * 成功返回：自定义消息
     *
     * @param data 响应数据
     * @param <T>  响应的数据类型
     * @return 响应数据
     */
    public static <T> Result<T> success(T data) {
        Result<T> result = new Result<>();
        result.setCode(ResCode.SUCCESS.getCode());
        result.setMessage(ResCode.SUCCESS.getMessage());
        result.setData(data);
        return result;
    }

    /**
     * 失败返回
     *
     * @param ResCode 响应码枚举
     * @param <T>     响应的数据类型
     * @return 响应数据
     */
    public static <T> Result<T> fail(ResCode ResCode) {
        Result<T> result = new Result<>();
        result.setCode(ResCode.getCode());
        result.setMessage(ResCode.getMessage());
        return result;
    }

    /**
     * 失败返回：自定义消息
     *
     * @param code    响应码
     * @param message 响应消息
     * @param <T>     响应的数据类型
     * @return 响应数据
     */
    public static <T> Result<T> fail(String code, String message) {
        Result<T> result = new Result<>();
        result.setCode(code);
        result.setMessage(message);
        return result;
    }
}

File: mrecord-rust/src/common/result.rs
//! 统一响应体
//!
//! 对应 Java: `com.dcz.mrecord.common.Result<T>`
//!
//! 注意：Rust 标准库已有 `Result` 类型，因此命名为 `ApiResponse` 避免冲突。

use axum::{Json, response::IntoResponse};
use serde::Serialize;

use super::res_code::ResCode;

/// 统一 API 响应结构
///
/// 所有接口返回 `{ "code": "...", "message": "...", "data": ... }` 格式。
#[derive(Serialize)]
pub struct ApiResponse<T: Serialize> {
    /// 响应码（如 "00000" 成功）
    pub code: String,
    /// 响应消息
    pub message: String,
    /// 响应数据，失败时为 null
    pub data: Option<T>,
}

impl<T: Serialize> ApiResponse<T> {
    /// 成功返回（带数据）
    pub fn success(data: T) -> Self {
        Self {
            code: ResCode::Success.code().to_string(),
            message: ResCode::Success.message().to_string(),
            data: Some(data),
        }
    }

    /// 成功返回（无数据）
    pub fn success_empty() -> ApiResponse<()> {
        ApiResponse {
            code: ResCode::Success.code().to_string(),
            message: ResCode::Success.message().to_string(),
            data: None,
        }
    }

    /// 失败返回（使用预定义响应码）
    pub fn fail(res_code: ResCode) -> Self {
        Self {
            code: res_code.code().to_string(),
            message: res_code.message().to_string(),
            data: None,
        }
    }

    /// 失败返回（自定义 code + message）
    pub fn fail_with(code: String, message: String) -> Self {
        Self {
            code,
            message,
            data: None,
        }
    }
}

impl<T: Serialize> IntoResponse for ApiResponse<T> {
    fn into_response(self) -> axum::response::Response {
        Json(self).into_response()
    }
}
